src/PubSubClient.cpp

```diff
@@ -12,93 +12,125 @@ PubSubClient::PubSubClient() {
     setServer(addr,port);
     setCallback(callback);
     setClient(client);
     setStream(stream);
+    this->buffer_size = MQTT_MAX_PACKET_SIZE;
+    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
 }
 
 PubSubClient::PubSubClient(uint8_t *ip, uint16_t port, Client& client) {
     this->_state = MQTT_DISCONNECTED;
     setServer(ip, port);
     setClient(client);
     this->stream = NULL;
+    this->buffer_size = MQTT_MAX_PACKET_SIZE;
+    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
 }
 PubSubClient::PubSubClient(uint8_t *ip, uint16_t port, Client& client, Stream& stream) {
     this->_state = MQTT_DISCONNECTED;
     setServer(ip,port);
     setClient(client);
     setStream(stream);
+    this->buffer_size = MQTT_MAX_PACKET_SIZE;
+    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
 }
 PubSubClient::PubSubClient(uint8_t *ip, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client) {
     this->_state = MQTT_DISCONNECTED;
     setServer(ip, port);
     setCallback(callback);
     setClient(client);
     this->stream = NULL;
+    this->buffer_size = MQTT_MAX_PACKET_SIZE;
+    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
 }
 PubSubClient::PubSubClient(uint8_t *ip, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client, Stream& stream) {
     this->_state = MQTT_DISCONNECTED;
     setServer(ip,port);
     setCallback(callback);
     setClient(client);
     setStream(stream);
+    this->buffer_size = MQTT_MAX_PACKET_SIZE;
+    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
 }
 
 PubSubClient::PubSubClient(const char* domain, uint16_t port, Client& client) {
     this->_state = MQTT_DISCONNECTED;
     setServer(domain,port);
     setClient(client);
     this->stream = NULL;
+    this->buffer_size = MQTT_MAX_PACKET_SIZE;
+    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
 }
 PubSubClient::PubSubClient(const char* domain, uint16_t port, Client& client, Stream& stream) {
     this->_state = MQTT_DISCONNECTED;
     setServer(domain,port);
     setClient(client);
     setStream(stream);
+    this->buffer_size = MQTT_MAX_PACKET_SIZE;
+    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
 }
 PubSubClient::PubSubClient(const char* domain, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client) {
     this->_state = MQTT_DISCONNECTED;
     setServer(domain,port);
     setCallback(callback);
     setClient(client);
     this->stream = NULL;
+    this->buffer_size = MQTT_MAX_PACKET_SIZE;
+    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
 }
 PubSubClient::PubSubClient(const char* domain, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client, Stream& stream) {
     this->_state = MQTT_DISCONNECTED;
     setServer(domain,port);
     setCallback(callback);
     setClient(client);
     setStream(stream);
+    this->buffer_size = MQTT_MAX_PACKET_SIZE;
+    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
+}
+
+PubSubClient::~PubSubClient() {
+  free(this->buffer);
```

Comment:
This should check if this->buffer is NULL or not since setBufferSize() can set it to NULL.

Reply:
Nope. according to the spec std::free() should treat a null pointer as a no-op.
from http://www.open-std.org/JTC1/SC22/wg14/www/docs/n1124.pdf :
> The free function causes the space pointed to by ptr to be deallocated, that is, made available for further allocation. If ptr is a null pointer, no action occurs.

(it makes sense because std::malloc returns null pointer on failure.)

Reply:
Ah, sweet, I haven't really coded C/C++ in ages and back then it was a no-no since freeing a null pointer would cause some environments to go belly up. :-)